codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1042 @@
+        // If we have many child channel we can optimize for the case when multiple call flush() in
+        // channelReadComplete(...) callbacks and only do it once as otherwise we will end-up with multiple
+        // write calls on the socket which is expensive.
+        try {
+            DefaultHttp2StreamChannel current = head;
+            while (current != null) {
+                DefaultHttp2StreamChannel childChannel = current;
+                if (childChannel.fireChannelReadPending) {
+                    // Clear early in case fireChildReadComplete() causes it to need to be re-processed
+                    childChannel.fireChannelReadPending = false;
+                    if (childChannel.fireChildReadComplete()) {
+                        flushNeeded = true;
+                    }
+                }
+                childChannel.next = null;
+                current = current.next;
+            }
+        } finally {
+            // We always flush as this is what Http2ConnectionHandler does for now.
+            // TODO: I think this is not really necessary and we should be able to optimize this in the future.
+            flush0(ctx);
         }
-        channel.pipeline().addLast(handler);
+    }
 
-        initOpts(channel, options);
-        initAttrs(channel, attrs);
+    // Allow to override for testing
+    void flush0(ChannelHandlerContext ctx) {
+        flush(ctx);
+    }
 
-        ChannelFuture future = group.register(channel);
-        // Handle any errors that occurred on the local thread while registering. Even though
-        // failures can happen after this point, they will be handled by the channel by closing the
-        // channel.
-        if (future.cause() != null) {
-            if (channel.isRegistered()) {
-                channel.close();
-            } else {
-                channel.unsafe().closeForcibly();
-            }
-        }
-        return future;
+    // Allow to override for testing
+    void onBytesConsumed(ChannelHandlerContext ctx, Http2FrameStream stream, int bytes) throws Http2Exception {
+        consumeBytes(stream.id(), bytes);
     }
 
-    @SuppressWarnings("unchecked")
-    private static void initOpts(Channel channel, Map<ChannelOption<?>, Object> opts) {
-        if (opts != null) {
-            for (Entry<ChannelOption<?>, Object> e: opts.entrySet()) {
-                try {
-                    if (!channel.config().setOption((ChannelOption<Object>) e.getKey(), e.getValue())) {
-                        logger.warn("Unknown channel option: " + e);
-                    }
-                } catch (Throwable t) {
-                    logger.warn("Failed to set a channel option: " + channel, t);
+    // Allow to extend for testing
+    static class Http2MultiplexCodecStream extends DefaultHttp2FrameStream {
+
+        DefaultHttp2StreamChannel channel;
+    }
+
+    // TODO: Handle writability changes due writing from outside the eventloop.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)